src/main/java/com/commercetools/sync/categories/CategorySync.java

```diff
@@ -395,35 +420,79 @@ private boolean isMissingCategory(@Nonnull final String categoryKey,
-                    if (createdChild.isPresent()) {
-                        final Category category = createdChild.get();
-                        final CategoryDraft categoryDraft =
-                            CategoryDraftBuilder.of(category)
-                                                .parent(createdCategory.toResourceIdentifier())
-                                                .build();
-                        categoryDraftsToUpdate.put(categoryDraft, category);
-                    } else {
-                        categoryKeysToFetch.add(childCategoryKey);
-                    }
+        unresolvedReferencesService.fetch(resolvableCategoryKeys,
+            CUSTOM_OBJECT_CATEGORY_CONTAINER_KEY,
+            WaitingToBeResolvedCategories.class)
+            .handle(ImmutablePair::new)
+            .thenApply(fetchResponse -> {
+                final List<CategoryDraft> readyToSync = new ArrayList<>();
+                final Set<? extends WaitingToBeResolved> waitingDrafts = fetchResponse.getKey();
+                final Throwable fetchException = fetchResponse.getValue();
+                if (fetchException != null) {
+                    final String errorMessage = format(UNRESOLVED_PARENT_REFERENCES_STORE_FETCH_FAILED,
+                        String.join(",", resolvableCategoryKeys.toString()));
+                    handleError(errorMessage, new SyncException(errorMessage, fetchException));
+                    return readyToSync;
                 }
-            }
-        });
+                //Each waitingdraft have only one parents, so we can sync the waitingdraft right away, because only
+                // waitingdraft with resolved categories was fetched
+                waitingDrafts.forEach(draft -> {
+                    final CategoryDraft categoryDraft = (CategoryDraft) draft.getWaitingDraft();
+                    readyToSync.add(categoryDraft);
+                });
+                return readyToSync;
+            })
+            .thenAccept(readyToSync -> {
+                if (!readyToSync.isEmpty()) {
+                    // process ready drafts
+                    final Set<CategoryDraft> readyToCreate = new HashSet<>();
+                    final Set<CategoryDraft> readyToUpate = new HashSet<>();
+                    readyToSync.stream().forEach(categoryDraft -> {
+                        referenceResolver.resolveReferences(categoryDraft)
+                            .thenAccept(referencesResolvedDraft -> {
+                                referencesResolvedDrafts.add(referencesResolvedDraft);
+                                if (keyToIdCache.containsKey(referencesResolvedDraft.getKey())) {
+                                    readyToUpate.add(referencesResolvedDraft);
+                                } else {
+                                    readyToCreate.add(referencesResolvedDraft);
+                                }
+                            }).toCompletableFuture().join();
+                    });
```

Comment:
can't you re-use `prepareDraftsForProcessing` method, or at least parts of it? seems that this is duplicated code.

Reply:
3bd67e05d